sycl/include/CL/sycl/handler.hpp

```diff
@@ -619,93 +592,109 @@ class handler {
+    kernel_parallel_for<NameT, KernelType, Dims>(KernelFunc);
 #else
     MNDRDesc.set(std::move(NumWorkItems));
-    StoreLambda<KernelName, KernelType, Dims>(std::move(KernelFunc));
+    StoreLambda<NameT, KernelType, Dims>(std::move(KernelFunc));
     MCGType = detail::CG::KERNEL;
 #endif
   }
 
-  // parallel_for version with a kernel represented as a functor + range that
-  // specifies global size only. Simply redirect to the lambda-based form of
-  // invocation, setting kernel name type to the functor type.
-  template <typename KernelType, int Dims>
-  void parallel_for(range<Dims> NumWorkItems, KernelType KernelFunc) {
-    parallel_for<KernelType, KernelType, Dims>(NumWorkItems, KernelFunc);
-  }
-
   // parallel_for version with a kernel represented as a lambda + range and
   // offset that specify global size and global offset correspondingly.
-  template <typename KernelName, typename KernelType, int Dims>
+  template <typename KernelName = std::nullptr_t, typename KernelType, int Dims>
   void parallel_for(range<Dims> NumWorkItems, id<Dims> WorkItemOffset,
                     KernelType KernelFunc) {
+    using NameT = typename csd::get_kernel_name_t<KernelName, KernelType>::name;
 #ifdef __SYCL_DEVICE_ONLY__
-    kernel_parallel_for<KernelName, KernelType, Dims>(KernelFunc);
+    kernel_parallel_for<NameT, KernelType, Dims>(KernelFunc);
 #else
     MNDRDesc.set(std::move(NumWorkItems), std::move(WorkItemOffset));
-    StoreLambda<KernelName, KernelType, Dims>(std::move(KernelFunc));
+    StoreLambda<NameT, KernelType, Dims>(std::move(KernelFunc));
     MCGType = detail::CG::KERNEL;
 #endif
   }
 
   // parallel_for version with a kernel represented as a lambda + nd_range that
   // specifies global, local sizes and offset.
-  template <typename KernelName, typename KernelType, int Dims>
+  template <typename KernelName = std::nullptr_t, typename KernelType, int Dims>
   void parallel_for(nd_range<Dims> ExecutionRange, KernelType KernelFunc) {
+    using NameT = typename csd::get_kernel_name_t<KernelName, KernelType>::name;
 #ifdef __SYCL_DEVICE_ONLY__
-    kernel_parallel_for<KernelName, KernelType, Dims>(KernelFunc);
+    kernel_parallel_for<NameT, KernelType, Dims>(KernelFunc);
 #else
     MNDRDesc.set(std::move(ExecutionRange));
-    StoreLambda<KernelName, KernelType, Dims>(std::move(KernelFunc));
+    StoreLambda<NameT, KernelType, Dims>(std::move(KernelFunc));
     MCGType = detail::CG::KERNEL;
 #endif
   }
 
-  // parallel_for version with a kernel represented as a functor + nd_range that
-  // specifies global, local sizes and offset. Simply redirect to the
-  // lambda-based form of invocation, setting kernel name type to the functor
-  // type.
-  template <typename KernelType, int Dims>
-  void parallel_for(nd_range<Dims> ExecutionRange, KernelType KernelFunc) {
-    parallel_for<KernelType, KernelType, Dims>(ExecutionRange, KernelFunc);
+  template <typename KernelName = std::nullptr_t, typename KernelType, int Dims>
```

Comment:
Possibly a reason behind it that I'm not seeing at the moment, but is there a reason the dimension is named "Dims" here and "dimensions" in kernel_parallel_for_work_group and in the other parallel_for_work_group interface (https://github.com/intel/llvm/pull/221/files#diff-234ddc8d342a3b24bb61d130262e9624R665). Perhaps worth renaming the other two that use "dimensions" to "Dims" to keep consistency.

Reply:
No particular reason. Will make it "Dims" - I agree, should be uniform.